compliance_checker/cf/cf.py

```diff
@@ -677,6 +707,90 @@ def _split_standard_name(self, standard_name):
         else:
             return standard_name, None
 
+    def _check_attr_type(self, attribute, attr_type, variable=None):
+        """
+        Check if an attribute `attr` is of the type `attr_type`. Upon getting
+        a data type of 'D', the attr must have the same data type as the
+        variable it is assigned to.
+
+        Attributes designated type 'S' must be of type `str`. 'N' require
+        numeric types, and 'D' requires the attribute type match the type
+        of the variable it is assigned to.
+
+        :param attribute: attribute to check
+        :param str attr_type: the correct type of the attribute
+        :param variable: if given, type should match attr
+        :rtype list
+        :return A two-element list that contains pass/fail status as a boolean and
+                a message string (or None if unset) as the second element.
+                The string will contain a format field to be filled with a
+                string describing the attribute.
+        """
+
+        if attr_type == 'S':
+            if not isinstance(attribute, basestring):
+                return [False,
+                        "{} must be a string".format('{}')]
+        else:
+            # if it's not a string, it should have a numpy dtype
+            underlying_dtype = getattr(attribute, 'dtype', None)
+
+            # TODO check for np.nan separately
+            if underlying_dtype is None:
+                return [False,
+                        "{} must be a numeric numpy datatype".format('{}')]
+
+            # both D and N should be some kind of numeric value
+            is_numeric = np.issubdtype(underlying_dtype, np.number)
+            if attr_type == 'N':
+                if not is_numeric:
+                    return [False,
+                            "{} must be numeric".format('{}')]
+            elif attr_type == 'D':
+                # TODO: handle edge case where variable is unset here
+                var_dtype = getattr(variable, 'dtype', None)
+                if (underlying_dtype != var_dtype):
+                    return [False,
+                            "{} must be numeric and must match {} dtype".format('{}', var_dtype)]
```

Comment:
If you are really sure you want to format returning a format string, you should use `{{}}` within the string and skip the `'{}'` format argument.

Reply:
See newest commit for changes